code/datums/abilities/critter/stomach_retreat.dm

```diff
@@ -0,0 +1,95 @@
+/datum/targetable/critter/stomach_retreat
+	name = "Retreat to Stomach"
+	desc = "Turn yourself inside out for shelter! Must be inside a disposal chute."
+	icon_state = "mimic_stomach_retreat"
+	cooldown = 120 SECONDS
+	cooldown_after_action = TRUE
+	needs_turf = FALSE
+	var/inside = FALSE
+	var/list/trap_whitelist = list(/obj/machinery/disposal, /obj/storage/)
+	var/obj/current_container = null
+	var/last_appearance = null
+
+	cast(atom/target)
+		. = ..()
+		if (inside)
+			switch(tgui_alert(holder.owner, "Leave yourself?", "Retreat to Stomach", list("Yes.", "No.")))
+				if ("Yes.")
+					deactivate()
+				if ("No.")
+					return TRUE
+		else
+			switch(tgui_alert(holder.owner, "Retreat into yourself to heal?", "Retreat to Stomach", list("Yes.", "No.")))
+				if ("Yes.")
+					var/turf/T = get_turf(holder.owner)
+					var/obj/target_container = holder.owner.loc
+					if (!T.z || isrestrictedz(T.z))
+						boutput(holder.owner, SPAN_ALERT("You are forbidden from using that here!"))
+						return TRUE
+					// Attempt entry via disposal machinery OR a disconnected disposal pipe
+					if (target_container.type in trap_whitelist)
```

Comment:
This needs to use `istypes`
(Was unable to stomach_retreat in `/obj/storage/closet` )